### DISCUSSION TITLE
2022 Road map

### DISCUSSION THREAD

**Discussion by gaborcsapo** [ARMmbed/DAPLink#886] [General]
Hey, thanks for all the hard work on the latest marge into master, the GCC support is greatly appreciated. I was wondering what's coming after this change, are there already big features in planning for the coming year?

**Comment by gaborcsapo** [comment]
Can you please elaborate on the CMSIS DAP 2.1 UART feature mentioned in https://github.com/ARMmbed/DAPLink/pull/870? Does this mean in that developers can have UART communication to the target over DAP commands, not just the USB port? 

We are planning to add an interactive shell over a new CDC endpoint to our stm32h7 HIC to allow easier interaction and control of our Daplink (read/write GPIOs, control peripheral, etc...). Do you see any potential conflict in your road map with this plan?

**Reply by flit** [reply]
That's right, the UART COM commands allow access to the UART via CMSIS-DAP. The CMSIS-DAP v2.1 documentation [has details about the commands](https://arm-software.github.io/CMSIS_5/develop/DAP/html/group__DAP__uart__gr.html).

Note that DAPLink already has support for this feature using vendor-specific commands. See [DAP_vendor.c](https://github.com/ARMmbed/DAPLink/blob/7839c6cef09eb1b3ae66e180450753e237559bd7/source/daplink/cmsis-dap/DAP_vendor.c#L100). We aren't planning to drop the vendor-specific UART commands any time soon since they are used heavily by micro:bit, but they might be deprecated (only once v2.1 UART COM commands are integrated, of course).

The main reason the UART COM commands aren't yet integrated is that the reference code requires CMSIS-UART drivers, and many of the HICs only have DAPLink's UART driver.

**Comment by mbrossard** [comment]
The roadmap will be what contributors want to work on (with some caveats, to keep the project useful and maintainable).

I can share a few of the topic areas / ideas that could see some progress next year(s)

### Disaggregation of DAPLink

There is a discussion in #868 about using parts of DAPLink for a different purpose. In a similar line of thought, could we make DAPLink in such a way that you would not need to fork it to add support for a HIC or a target? On those aspects, selecting targets in post-processing is probably the closest to become a reality: a lot of the pieces are in place, it just has not be put in practice yet. The idea is that you build one interface firmware for each HIC create, one object per target and you can from those assemble any combinations with a script.

### Leveraging the CMSIS ecosystem

With some of the changes in CMSIS-DAP v2.1 there is now an additional benefit to using UART CMSIS Drivers: previous version enable SWO support through them, CMSIS-DAP v2.1 add support for UART communication through DAP commands to the list. There are other potential opportunities to gain from the CMSIS ecosystem and new initiatives like Open-CMSIS-Pack. A lot of the code in HICs source code and target support information either comes from or is present in CMSIS packs.

@flit created an adapter between DAPLink's CDC support and a CMSIS UART Driver for the `lpc55xx` HIC, so it uses the same driver for CDC/ACM and SWO ports. This effort shows the way that could be followed by other HICs where `uart.c`, a semi-custom driver, is replaced by a generic driver (usually of much better quality) and in the process add support for SWO where possible.

### Clean-ups, updates and testing

I discovered a lot of the DAPLink code base by porting it to GCC and armclang. I noticed the HIC code contains a lot of code from old CMSIS packs or SDKs. Some of the initialization (clocks, pin-mux, GPIO, etc.) is not always consistent between HICs, which is something I have tried to improve over the past year. It would make sense to converge all HICs to a similar organization and using the latest upstream supported resources (which are usually cleaner with some bugs fixed).

The `test` contains a test suite for DAPLink that is already very useful. There are still potential for increased coverage of the feature set of DAPLink (e.g. WebUSB, SWO, etc.). One thing that is currently also missing from it are performance metrics for SWD, UART, Flash speed.

### Miscellaneous

In addition to some of the tasks mentioned above, I am personally thinking about working on some additions (those ideas are not exclusive, you can help 😄):
- I have a few candidates I am thinking about to add as HICs, I would be interested to know what people would like to see before I share my thoughts.
- `UF2` file support
- Find a way to add support for soft-reset in a clean way.

**Reply by raveslave** [reply]
curious what's available via webusb today?   
it would be nice with something similar to how nordic presents it in their nrf connect - programmer
